applybn/feature_generation/bn_feature_generator.py

```diff
@@ -0,0 +1,158 @@
+import random
+import numpy as np
+import pandas as pd
+from pgmpy.models import BayesianNetwork
+from pgmpy.estimators import K2Score, BayesianEstimator, BicScore
+from pgmpy.inference import VariableElimination
+
+
+class BNFeatureGenerator:
+    def __init__(self, known_structure=None):
+        self.known_structure = known_structure
+        self.bn = None
+        self.variables = None
+        self.num_classes = None
+
+    def fit(self, data, target=None, black_list=None):
+        if not isinstance(data, pd.DataFrame):
+            data = pd.DataFrame(data)
+
+        self.variables = data.columns.tolist()
+
+        if target is not None:
+            if not isinstance(target, pd.Series):
+                target = pd.Series(target)
+            self.num_classes = len(target.unique())
+
+        if self.known_structure:
+            self.bn = BayesianNetwork(self.known_structure)
+            self.bn.fit(data, estimator=BayesianEstimator)
```

Comment:
When the method is provided a known_structure, it immediately calls bn.fit(...) using BayesianEstimator. It might be helpful to explicitly catch and log any exceptions from pgmpy’s fitting process to inform users if the known structure is invalid or the data is insufficient.

```suggestion
  try:
      self.bn.fit(data, estimator=BayesianEstimator)
  except Exception as e:
      # Log details about the fitting error
      logger.exception(f"Error fitting BayesianNetwork with known structure: {e}")
      raise
```